lib/lightning_web/live/project_live/new_collaborator_component.ex

```diff
@@ -16,35 +19,74 @@ defmodule LightningWeb.ProjectLive.NewCollaboratorComponent do
     {:ok,
      socket
      |> assign(assigns)
-     |> assign(collaborators: collaborators, changeset: changeset)}
+     |> assign(collaborators: collaborators, changeset: changeset, error: nil)}
   end
 
   @impl true
   def handle_event("validate", %{"project" => params}, socket) do
-    changeset =
-      Collaborators.changeset(socket.assigns.collaborators, params)
+    socket =
+      assign(socket,
+        changeset: Collaborators.changeset(socket.assigns.collaborators, params)
+      )
 
-    {:noreply, assign(socket, changeset: changeset)}
+    with :ok <- limit_adding_users(socket, params) do
+      {:noreply, assign(socket, error: nil)}
+    end
   end
 
-  def handle_event(
-        "add_collaborators",
-        %{"project" => params},
-        %{assigns: assigns} = socket
-      ) do
-    with {:ok, project_users} <-
-           Collaborators.prepare_for_insertion(
-             assigns.collaborators,
-             params,
-             assigns.project_users
-           ),
-         {:ok, _} <-
-           Projects.add_project_users(assigns.project, project_users) do
+  def handle_event("add_collaborators", %{"project" => params}, socket) do
+    with :ok <- limit_adding_users(socket, params),
+         {:ok, project_users} <- prepare_for_insertion(socket, params),
+         {:ok, _project} <- add_project_users(socket, project_users) do
       send(self(), :collaborators_updated)
       {:noreply, socket}
-    else
+    end
+  end
+
+  defp prepare_for_insertion(%{assigns: assigns} = socket, params) do
+    case Collaborators.prepare_for_insertion(
+           assigns.collaborators,
+           params,
+           assigns.project_users
+         ) do
+      {:ok, project_users} ->
+        {:ok, project_users}
+
       {:error, changeset} ->
         {:noreply, assign(socket, changeset: changeset)}
     end
   end
+
+  defp add_project_users(socket, project_users) do
+    case Projects.add_project_users(socket.assigns.project, project_users) do
+      {:ok, project} ->
+        {:ok, project}
+
+      {:error, changeset} ->
+        {:noreply, assign(socket, changeset: changeset)}
+    end
+  end
+
+  defp limit_adding_users(socket, params) do
+    changeset =
+      Collaborators.changeset(socket.assigns.collaborators, params)
+
+    project_users = Ecto.Changeset.get_embed(changeset, :collaborators)
+
+    case UsageLimiter.limit_action(
+           %Action{
+             type: :new_user,
+             amount: Enum.count(project_users)
+           },
+           %Context{
+             project_id: socket.assigns.project.id
+           }
+         ) do
+      :ok ->
+        :ok
+
+      {:error, _reason, %{text: error_text}} ->
+        {:noreply, assign(socket, error: error_text)}
+    end
+  end
```

Comment:
Going forward, we should not be putting changesets functions especially when calls to `UsageLimiter` are needed inside LiveViews.

This stuff should be abstracted into `Lightning` and the use of `UsageLimiter` should be hidden inside a module dedicated to handling validation and calling the limiter.